torch/_tensor.py

```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
             else:
                 return _convert(ret, cls)
 
+    def __dlpack__(self, stream=None):
+        """
+        Creates a dlpack capsule of the current tensor to
+        be exported to other libraries.
+
+        This function will be called from the `from_dlpack` method
+        of the consumer library.
+
+        Args:
+            stream (integer or None): A Python integer representing a pointer
+            to a stream (CUDA or ROCm). `stream` is provided by the consumer
+            to the producer to instruct the producer to ensure that operations
+            can safely be performed on the array.
+            The pointer must be a positive integer or
+            -1 . If stream is -1 , the value may be used by the consumer to
+            signal "producer must not perform any synchronization. Optional.
+        """
+        if has_torch_function_unary(self):
+            return handle_torch_function(Tensor.__dlpack__, (self,), self, stream)
+        if stream is not None and type(stream) is not int:
+            # Stream pointers in CUDA/ROCm are uniquely numbered and can
+            # be retrieved from their integer value.
+            raise TypeError('stream must be ``int`` or ``none``')
+        elif stream is not None and stream != -1:
+            if self.device.type == 'cuda':
+                stream = torch.cuda.streams.ExternalStream(stream)
+                # Only synchronize on different streams
+                if stream != torch.cuda.current_stream:
+                    event = torch.cuda.Event()
+                    event.record(torch.cuda.current_stream())
+                    stream.wait_event(event)
+        return torch.to_dlpack(self)
```

Comment:
So if an integer is provided this uses that integer to acquire an ExternalStream, then syncs that external stream with the current stream, but then the `to_dlpack` call discards the stream operation?

What is this intended to accomplish?

Reply:
This ensures that the data will be copied at the right moment by forcing an order between the external library stream and pytorch stream.

https://github.com/dmlc/dlpack/issues/57#issuecomment-753696812

Reply:
OK I think I understand this now because the `to_dlpack` operation is not actually a CUDA operation since the memory is shared.